lua-api-crates/table-funcs/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+// can be useful in some cases
+fn merge<'lua>(
+    lua: &'lua Lua,
+    (array_of_tables, keep_first): (Vec<Table<'lua>>, Option<bool>),
+) -> mlua::Result<Table<'lua>> {
+    let mut tbl_vec: Vec<(LuaValue, LuaValue)> = vec![];
+    for table in array_of_tables {
+        for pair in table.pairs::<LuaValue, LuaValue>() {
+            let (key, value) = pair.map_err(mlua::Error::external)?;
+            tbl_vec.push((key, value));
+        }
+    }
+    let tbl_len = tbl_vec.len();
+    // note we might allocate a bit too much here, but in many use cases we will be correct
+    let tbl: Table<'lua> = lua
+        .create_table_with_capacity(0, tbl_len)
+        .map_err(mlua::Error::external)?;
+
+    let keep_first = match keep_first {
+        Some(b) => b,
+        None => false, // default behavior is to keep_last set value
+    };
+    for (key, value) in tbl_vec {
+        // Note that we override previously set key values if we have
+        // the same key showing up more than once
+        if keep_first {
+            if !tbl
+                .contains_key(key.clone())
+                .map_err(mlua::Error::external)?
+            {
+                tbl.set(key, value).map_err(mlua::Error::external)?;
+            }
+        } else {
+            tbl.set(key, value).map_err(mlua::Error::external)?;
+        }
+    }
+    Ok(tbl)
+}
+
+fn clone<'lua>(_: &'lua Lua, table: Table<'lua>) -> mlua::Result<Table<'lua>> {
+    Ok(table.clone())
+}
+
+fn flatten<'lua>(lua: &'lua Lua, arrays: Vec<LuaValue<'lua>>) -> mlua::Result<Vec<LuaValue<'lua>>> {
+    let mut flat_vec: Vec<LuaValue> = vec![];
+    for item in arrays {
+        match item {
+            LuaValue::Table(tbl) => {
+                let tbl_as_vec = tbl.sequence_values().filter_map(|x| x.ok()).collect();
+                let flat = flatten(lua, tbl_as_vec)?;
+                for j in flat {
+                    flat_vec.push(j);
+                }
+            }
+            LuaValue::Nil => (),
+            LuaValue::Thread(_) => (),
+            LuaValue::Error(err) => {
+                return Err(mlua::Error::external(err));
+            }
+            other => {
+                flat_vec.push(other);
+            }
+        }
+    }
+    Ok(flat_vec)
+}
+
+fn length<'lua>(_: &'lua Lua, table: Table<'lua>) -> mlua::Result<Integer> {
+    // note that # only works correctly on arrays in Lua
+    let mut len: i64 = 0;
+    for _ in table.pairs::<LuaValue, LuaValue>() {
+        len += 1
+    }
+    Ok(len)
+}
+
+fn has_key<'lua>(_: &'lua Lua, (table, key): (Table<'lua>, LuaValue)) -> mlua::Result<bool> {
+    Ok(table.contains_key(key).map_err(mlua::Error::external)?)
+}
+
+fn has_value<'lua>(_: &'lua Lua, (table, value): (Table<'lua>, LuaValue)) -> mlua::Result<bool> {
+    for pair in table.pairs::<LuaValue, LuaValue>() {
+        let (_, tbl_value) = pair.map_err(mlua::Error::external)?;
```

Comment:
Is the error here really an external error?  Does this work?
I think the same question applies to a number of other calls in this module; in general, errors returned from the mlua crate are already `mlua::Result` and shouldn't need to be mapped.

```suggestion
        let (_, tbl_value) = pair?;
```

Reply:
Oh, I didn't think about that. I will try to cleanup all the errors after going through the other comments.